cannon/mipsevm/multithreaded/mips.go

```diff
@@ -0,0 +1,318 @@
+				v0 = exec.SysErrorSignal
+				v1 = exec.MipsEAGAIN
+			} else {
+				thread.FutexVal = a2
+				if a3 == 0 {
+					thread.FutexTimeoutStep = exec.FutexNoTimeout
+				} else {
+					thread.FutexTimeoutStep = m.state.Step + exec.FutexTimeoutSteps
+				}
+				// Leave cpu scalars as-is. This instruction will be completed by `onWaitComplete`
+				return nil
+			}
+		case exec.FutexWakePrivate:
+			// Trigger thread traversal starting from the left stack until we find one waiting on the wakeup
+			// address
+			m.state.Wakeup = a0
+			// Don't indicate to the program that we've woken up a waiting thread, as there are no guarantees.
+			// The woken up thread should indicate this in userspace.
+			v0 = 0
+			v1 = 0
+			exec.HandleSyscallUpdates(&thread.Cpu, &thread.Registers, v0, v1)
+			m.preemptThread(thread, true)
+			m.state.TraverseRight = len(m.state.LeftThreadStack) == 0
+			return nil
+		default:
+			v0 = exec.SysErrorSignal
+			v1 = exec.MipsEINVAL
+		}
+	case exec.SysSchedYield, exec.SysNanosleep:
+		v0 = 0
+		v1 = 0
+		exec.HandleSyscallUpdates(&thread.Cpu, &thread.Registers, v0, v1)
+		m.preemptThread(thread, true)
+		return nil
+	case exec.SysOpen:
+		v0 = exec.SysErrorSignal
+		v1 = exec.MipsEBADF
+	case exec.SysGetAffinity:
+	case exec.SysMadvise:
+	case exec.SysRtSigprocmask:
+	case exec.SysSigaltstack:
+	case exec.SysRtSigaction:
+	case exec.SysPrlimit64:
+	case exec.SysClose:
+	case exec.SysPread64:
+	case exec.SysFstat64:
+	case exec.SysOpenAt:
+	case exec.SysReadlink:
+	case exec.SysReadlinkAt:
+	case exec.SysIoctl:
+	case exec.SysEpollCreate1:
+	case exec.SysPipe2:
+	case exec.SysEpollCtl:
+	case exec.SysEpollPwait:
+	case exec.SysGetRandom:
+	case exec.SysUname:
+	case exec.SysStat64:
+	case exec.SysGetuid:
+	case exec.SysGetgid:
+	case exec.SysLlseek:
+	case exec.SysMinCore:
+	case exec.SysTgkill:
+	case exec.SysSetITimer:
+	case exec.SysTimerCreate:
+	case exec.SysTimerSetTime:
+	case exec.SysTimerDelete:
+	case exec.SysClockGetTime:
+	default:
+		m.Traceback()
+		panic(fmt.Sprintf("unrecognized syscall: %d", syscallNum))
+	}
+
+	exec.HandleSyscallUpdates(&thread.Cpu, &thread.Registers, v0, v1)
+	return nil
+}
+
+func (m *InstrumentedState) mipsStep() error {
+	if m.state.Exited {
+		return nil
+	}
+	m.state.Step += 1
+
+	// If we've completed traversing both stacks for the wakeup check
+	if m.state.Wakeup != exec.FutexEmptyAddr && m.state.TraverseRight && len(m.state.RightThreadStack) == 0 {
+		m.state.TraverseRight = false
+		m.state.Wakeup = exec.FutexEmptyAddr
+		return nil
+	}
+
+	thread := m.state.getCurrentThread()
+	if thread.Exited {
+		m.popThread()
+		m.stackTracker.DropThread(thread.ThreadId)
+		return nil
+	}
+
+	// Search for the first thread blocked by the wakeup call, if wakeup is set
+	// Don't allow regular execution until we resolved if we have woken up any thread.
+	if m.state.Wakeup != exec.FutexEmptyAddr && m.state.Wakeup != thread.FutexAddr {
+		m.preemptThread(thread, !m.state.TraverseRight)
```

Comment:
Why don't we update direction when traversing right? If we've finished traversing right, then shouldn't we swap directions?

Reply:
This is so on the next step after preempting the last thread in the right stack, we will hit the terminating condition for the wakeup traversal and clear the wakeup [here](https://github.com/ethereum-optimism/optimism/blob/b0206187e7183fbbae99a6a11e879cf2520e43ad/cannon/mipsevm/multithreaded/mips.go#L184).

Reply:
reworking the wakeup traversal logic here to avoid this special case